### PR TITLE
check resource requirements and accessmodes

### PR DESCRIPTION
**Pull request by zouyee**
**What this PR does / why we need it**:
Check resource requirements and accessmodes, update if necessary

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```Complete TODO: Check resource requirements and accessmodes, update if necessary

```

### CODE REVIEW

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 1]
```diff
@@ -193,7 +194,31 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
 			errs = append(errs, fmt.Errorf("Failed to retrieve PVC %s: %s", claim.Name, err))
 			spc.recordClaimEvent("create", set, pod, &claim, err)
 		}
-		// TODO: Check resource requirements and accessmodes, update if necessary
+		// Check resource requirements and accessmodes, update if necessary
+		if err == nil {
```
Comment (dixudx) [inline pkg/controller/statefulset/stateful_pod_control.go]: I'd prefer to take this as another switch case.

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 2]
```diff
@@ -193,7 +194,31 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
...
+			change := false
```
Comment (dixudx) [inline pkg/controller/statefulset/stateful_pod_control.go]: `s/change/changed` ?

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 3]
```diff
@@ -193,7 +194,31 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
...
+			if len(claim.Spec.AccessModes) > 0 {
```
Comment (dixudx) [inline pkg/controller/statefulset/stateful_pod_control.go]: What if `len(claim.Spec.AccessModes)=0` and `len(currentClaim.Spec.AccessModes)>0` ?

Need to adjust the conditions.

Comment (zouyee) [reply pkg/controller/statefulset/stateful_pod_control.go]: We only check the condition of `len(claim.Spec.AccessModes)>0`, if the length is zero, we do nothing.

**Comment by zouyee** [comment]
ping @smarterclayton， would you like to review it？ thank you!

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 4]
```diff
@@ -192,8 +193,33 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
...
+					changed = true
+					currentClaim.Spec.AccessModes = claim.Spec.AccessModes
+				} else if !volume.AccessModesContainedInAll(claim.Spec.AccessModes, currentClaim.Spec.AccessModes) {
+					changed = true
+					currentClaim.Spec.AccessModes = claim.Spec.AccessModes
+				}
+				if storage, ok := claim.Spec.Resources.Requests[v1.ResourceStorage]; ok {
+					if currentStorage, check := currentClaim.Spec.Resources.Requests[v1.ResourceStorage]; check && storage.Cmp(currentStorage) != 0 {
+						changed = true
+						currentClaim.Spec.Resources.Requests[v1.ResourceStorage] = claim.Spec.Resources.Requests[v1.ResourceStorage]
```
Comment (dhilipkumars) [inline pkg/controller/statefulset/stateful_pod_control.go]: why cant we simply assign ?
```
currentClaim.Spec.Resources.Requests[v1.ResourceStorage] = storage
```

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 5]
```diff
@@ -192,8 +193,33 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
...
+			// When the length of AccessModes is zero, we do nothing.
+			if len(claim.Spec.AccessModes) > 0 {
+				if len(claim.Spec.AccessModes) != len(currentClaim.Spec.AccessModes) {
+					changed = true
+					currentClaim.Spec.AccessModes = claim.Spec.AccessModes
+				} else if !volume.AccessModesContainedInAll(claim.Spec.AccessModes, currentClaim.Spec.AccessModes) {
+					changed = true
+					currentClaim.Spec.AccessModes = claim.Spec.AccessModes
+				}
+				if storage, ok := claim.Spec.Resources.Requests[v1.ResourceStorage]; ok {
```
Comment (dhilipkumars) [inline pkg/controller/statefulset/stateful_pod_control.go]: could we re-structure it like below, i feel its more readable
```
claimStorage, claimOk := claim.Spec.Resources.Requests[v1.ResourceStorage]; 
currentStorage, currentOk := currentClaim.Spec.Resources.Requests[v1.ResourceStorage]; 

if claimOk  && currentOk && claimStorage.Cmp(currentStorage) != 0 {
```

**Review by dhilipkumars — CHANGES_REQUESTED** [review]
Should we not add UT for these newly added code?

**Pull request closed without merging**